File: src/main/java/work/xeltica/craft/core/handlers/EbiPowerHandler.java
package work.xeltica.craft.core.handlers;
import java.util.Date;
import java.util.HashSet;
import java.util.Random;

import org.bukkit.Material;
import org.bukkit.Sound;
import org.bukkit.SoundCategory;
import org.bukkit.Tag;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Ageable;
import org.bukkit.entity.Cat;
import org.bukkit.entity.Hoglin;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Monster;
import org.bukkit.entity.Ocelot;
import org.bukkit.entity.Player;
import org.bukkit.entity.SkeletonHorse;
import org.bukkit.entity.Tameable;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.entity.EntityBreedEvent;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDeathEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.Damageable;

import net.kyori.adventure.text.Component;
import work.xeltica.craft.core.events.RealTimeNewDayEvent;
import work.xeltica.craft.core.models.Hint;
import work.xeltica.craft.core.models.PlayerDataKey;
import work.xeltica.craft.core.stores.EbiPowerStore;
import work.xeltica.craft.core.stores.HintStore;
import work.xeltica.craft.core.stores.MobEPStore;
import work.xeltica.craft.core.stores.PlayerStore;

/**
 * エビパワー関連のイベントハンドラをまとめています。
 * @author Xeltica
 */
public class EbiPowerHandler implements Listener{
    public EbiPowerHandler() {
        // エビパワーが貯まらないワールドのリストを構築
        // TODO: 設定ファイルに移す
        epBlackList.add("hub2");
        epBlackList.add("art");
        epBlackList.add("sandbox2");
        epBlackList.add("pvp");
        epBlackList.add("hub_dev");

        crops.addAll(Tag.CROPS.getValues());

        breakBonusList.addAll(Tag.BASE_STONE_OVERWORLD.getValues());
        breakBonusList.addAll(Tag.BASE_STONE_NETHER.getValues());
        breakBonusList.addAll(Tag.ICE.getValues());
        breakBonusList.addAll(Tag.DIRT.getValues());
        breakBonusList.addAll(Tag.SAND.getValues());

        breakBonusList.addAll(Tag.COAL_ORES.getValues());
        breakBonusList.addAll(Tag.IRON_ORES.getValues());
        breakBonusList.addAll(Tag.COPPER_ORES.getValues());
        breakBonusList.addAll(Tag.GOLD_ORES.getValues());
        breakBonusList.addAll(Tag.REDSTONE_ORES.getValues());
        breakBonusList.addAll(Tag.EMERALD_ORES.getValues());
        breakBonusList.addAll(Tag.LAPIS_ORES.getValues());
        breakBonusList.addAll(Tag.DIAMOND_ORES.getValues());
        breakBonusList.addAll(Tag.LOGS.getValues());

        breakBonusList.add(Material.NETHER_QUARTZ_ORE);
        breakBonusList.add(Material.OBSIDIAN);
        breakBonusList.add(Material.ANCIENT_DEBRIS);
    }

    @EventHandler(priority = EventPriority.NORMAL)
    public void onPlayerDamageFrailCreatures(EntityDamageByEntityEvent e) {
        if (e.getDamager() instanceof Player killer && e.getEntity() instanceof LivingEntity victim) {
            if (playerIsInBlacklisted(killer)) return;
            // スポナーは対象外
            if (victim.fromMobSpawner()) return;
            // TODO: TT登録可能にしてその中のキルは対象外にする

            if (victim instanceof Cat || victim instanceof Ocelot) {
                store().tryTake(killer, 100);
                notification(killer, "可愛い可愛いネコちゃんを殴るなんて！100EPを失った。");
                killer.playSound(killer.getLocation(), Sound.ENTITY_ZOMBIE_VILLAGER_CURE, SoundCategory.PLAYERS, 0.7f, 0.5f);
                HintStore.getInstance().achieve(killer, Hint.VIOLENCE_CAT);
            } else if (victim instanceof Tameable pet && pet.isTamed() && !(victim instanceof SkeletonHorse)) {
                store().tryTake(killer, 10);
                notification(killer, "ペットを殴るなんて！10EPを失った。");
                killer.playSound(killer.getLocation(), Sound.ENTITY_ZOMBIE_VILLAGER_CURE, SoundCategory.PLAYERS, 0.7f, 0.5f);
                HintStore.getInstance().achieve(killer, Hint.VIOLENCE_PET);
            } else if (victim instanceof Ageable man && !(victim instanceof Monster) && !(victim instanceof Hoglin) && !man.isAdult()) {
                store().tryTake(killer, 10);
                notification(killer, "子供を殴るなんて！10EPを失った。");
                killer.playSound(killer.getLocation(), Sound.ENTITY_ZOMBIE_VILLAGER_CURE, SoundCategory.PLAYERS, 0.7f, 0.5f);
                HintStore.getInstance().achieve(killer, Hint.VIOLENCE_CHILD);
            }
        }
    }
    
    @EventHandler
    public void onPlayerKillMobs(EntityDeathEvent e) {
        final var victim = e.getEntity();
        final var killer = e.getEntity().getKiller();
        if (killer != null) {
            if (playerIsInBlacklisted(killer)) return;
            // don't kill cats
            if (victim instanceof Cat || victim instanceof Ocelot) return;
            // don't kill tamed pets
            if (victim instanceof Tameable pet && pet.isTamed() && !(victim instanceof SkeletonHorse)) return;
            // don't kill non-monster children
            if (victim instanceof Ageable man && !(victim instanceof Monster) && !(victim instanceof Hoglin) && !man.isAdult()) return;
            // ignore creatures from spawner
            if (victim.fromMobSpawner()) return;
            var ep = "nightmare2".equals(killer.getWorld().getName()) ? MobEPStore.getInstance().getMobDropEP(victim, e) : 6;
            final var buff = getMobDropBonus(killer.getInventory().getItemInMainHand()) * 4;
            ep += (buff > 0 ? random.nextInt(buff) : 0);
            if (ep > 0) {
                store().tryGive(killer, ep);
                HintStore.getInstance().achieve(killer, Hint.KILL_MOB_AND_EARN_MONEY);
            }
        }
    }

    @EventHandler
    public void onPlayerLoggedIn(PlayerJoinEvent e) {
        final var now = new Date();
        final var ps = PlayerStore.getInstance();
        final var record = ps.open(e.getPlayer());
        final var prev = new Date(record.getLong(PlayerDataKey.LAST_JOINED, now.getTime()));
        if (prev.getYear() != now.getYear() && prev.getMonth() != now.getMonth() && prev.getDate() != now.getDate()) {
            store().tryGive(e.getPlayer(), LOGIN_BONUS_POWER);
            notification(e.getPlayer(), "ログボ達成！" + LOGIN_BONUS_POWER + "EPを獲得。");
        }
        record.set(PlayerDataKey.LAST_JOINED, now.getTime());
    }

    @EventHandler
    public void onHarvestCrops(BlockBreakEvent e) {
        final var p = e.getPlayer();
        if (playerIsInBlacklisted(p)) return;
        if (e.getBlock().getBlockData() instanceof org.bukkit.block.data.Ageable a && a.getAge() == a.getMaximumAge()) {
            final var tool = p.getInventory().getItemInMainHand();
            final var bonus = getBlockDropBonus(tool);
            final var power = (1 + bonus) * HARVEST_POWER_MULTIPLIER;
            store().tryGive(p, power);
            HintStore.getInstance().achieve(p, Hint.HARVEST_AND_EARN_MONEY);
            // もし幸運ボーナスがあれば30%の確率で耐久が減っていく
            if (bonus > 0 && random.nextInt(100) < 30) {
                tool.editMeta(meta -> {
                    if (meta instanceof Damageable toolMeta) {
                        toolMeta.setDamage(toolMeta.getDamage() + 1);
                    }
                });
            }
        }
    }

    @EventHandler
    public void onBreedEntities(EntityBreedEvent e) {
        if (e.getBreeder() instanceof Player player) {
            if (playerIsInBlacklisted(player)) return;
            EbiPowerStore.getInstance().tryGive(player, 2);
            HintStore.getInstance().achieve(player, Hint.BREED_AND_EARN_MONEY);
        }
    }

    @EventHandler
    public void onMineBlocks(BlockBreakEvent e) {
        if (!breakBonusList.contains(e.getBlock().getType())) return;
        if (playerIsInBlacklisted(e.getPlayer())) return;

        final var record = PlayerStore.getInstance().open(e.getPlayer());
        final var brokenBlocksCount = record.getInt(PlayerDataKey.BROKEN_BLOCKS_COUNT);
        // リミット超えていたら対象外
        if (brokenBlocksCount > BREAK_BLOCK_BONUS_LIMIT) return;

        // 回収できなかったら対象外
        if (!e.isDropItems()) return;

        // シルクタッチは対象外
        final var item = e.getPlayer().getInventory().getItemInMainHand();
        if (item.containsEnchantment(Enchantment.SILK_TOUCH)) return;

        record.set(PlayerDataKey.BROKEN_BLOCKS_COUNT, brokenBlocksCount + 1);

        if (brokenBlocksCount + 1 >= BREAK_BLOCK_BONUS_LIMIT) {
            HintStore.getInstance().achieve(e.getPlayer(), Hint.MINERS_DREAM);
        }
        store().tryGive(e.getPlayer(), 1);
    }

    @EventHandler
    public void onNewDayToResetBrokenBlocksCount(RealTimeNewDayEvent e) {
        PlayerStore.getInstance().openAll().forEach(record -> record.set(PlayerDataKey.BROKEN_BLOCKS_COUNT, 0));
    }

    private void notification(Player p, String mes) {
        p.sendActionBar(Component.text(mes));
        // p.playSound(p.getLocation(), Sound.ENTITY_PLAYER_LEVELUP, SoundCategory.PLAYERS, 1, 2);
    }

    private boolean playerIsInBlacklisted(Player p) {
        final var wName = p.getWorld().getName();
        return epBlackList.contains(wName);
    }

    private EbiPowerStore store() {
        return EbiPowerStore.getInstance();
    }

    private int getMobDropBonus(ItemStack stack) {
        if (stack == null) return 0;
        return stack.getEnchantmentLevel(Enchantment.LOOT_BONUS_MOBS);
    }

    private int getBlockDropBonus(ItemStack stack) {
        if (stack == null) return 0;
        return stack.getEnchantmentLevel(Enchantment.LOOT_BONUS_BLOCKS);
    }

    private final HashSet<String> epBlackList = new HashSet<>();

    private final HashSet<Material> crops = new HashSet<>();
    private final HashSet<Material> breakBonusList = new HashSet<>();
    private final HashSet<Material> placeBonusList = new HashSet<>();

    private static final int HARVEST_POWER_MULTIPLIER = 1;
    private static final int LOGIN_BONUS_POWER = 50;
    public static final int BREAK_BLOCK_BONUS_LIMIT = 4000;

    private static final Random random = new Random();
}

File: src/main/java/work/xeltica/craft/core/modules/ebipower/EbiPowerHandler.kt
package work.xeltica.craft.core.modules.ebipower

import com.destroystokyo.paper.MaterialTags
import net.kyori.adventure.text.Component
import org.bukkit.GameMode
import org.bukkit.Material
import org.bukkit.Sound
import org.bukkit.SoundCategory
import org.bukkit.Tag
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.Ageable
import org.bukkit.entity.Cat
import org.bukkit.entity.Hoglin
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Monster
import org.bukkit.entity.Ocelot
import org.bukkit.entity.Player
import org.bukkit.entity.SkeletonHorse
import org.bukkit.entity.Slime
import org.bukkit.entity.Tameable
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.entity.CreatureSpawnEvent
import org.bukkit.event.entity.EntityBreedEvent
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.entity.EntityDeathEvent
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.Damageable
import work.xeltica.craft.core.api.events.RealTimeNewDayEvent
import work.xeltica.craft.core.api.playerStore.PlayerStore
import work.xeltica.craft.core.hooks.CitizensHook.isCitizensNpc
import work.xeltica.craft.core.modules.hint.Hint
import work.xeltica.craft.core.modules.hint.HintModule
import work.xeltica.craft.core.modules.transferPlayerData.TransferPlayerDataEvent
import work.xeltica.craft.core.modules.world.WorldModule
import java.util.Random

/**
 * エビパワー関連のイベントハンドラをまとめています。
 * @author raink1208
 */
class EbiPowerHandler : Listener {
    companion object {
        private const val HARVEST_POWER_MULTIPLIER = 1
        const val BREAK_BLOCK_BONUS_LIMIT = 4000
    }

    private val crops = HashSet<Material>()
    private val breakBonusList = HashSet<Material>()

    init {
        crops.addAll(Tag.CROPS.values)

        breakBonusList.addAll(Tag.BASE_STONE_OVERWORLD.values)
        breakBonusList.addAll(Tag.BASE_STONE_NETHER.values)
        breakBonusList.addAll(Tag.ICE.values)
        breakBonusList.addAll(Tag.DIRT.values)
        breakBonusList.addAll(Tag.SAND.values)

        breakBonusList.addAll(Tag.COAL_ORES.values)
        breakBonusList.addAll(Tag.IRON_ORES.values)
        breakBonusList.addAll(Tag.COPPER_ORES.values)
        breakBonusList.addAll(Tag.GOLD_ORES.values)
        breakBonusList.addAll(Tag.REDSTONE_ORES.values)
        breakBonusList.addAll(Tag.EMERALD_ORES.values)
        breakBonusList.addAll(Tag.LAPIS_ORES.values)
        breakBonusList.addAll(Tag.DIAMOND_ORES.values)
        breakBonusList.addAll(Tag.LOGS.values)
        breakBonusList.addAll(MaterialTags.TERRACOTTA.values)

        breakBonusList.add(Material.NETHER_QUARTZ_ORE)
        breakBonusList.add(Material.OBSIDIAN)
        breakBonusList.add(Material.ANCIENT_DEBRIS)
        breakBonusList.add(Material.SANDSTONE)
        breakBonusList.add(Material.DEEPSLATE)
        breakBonusList.add(Material.KELP_PLANT)
        breakBonusList.add(Material.DRIPSTONE_BLOCK)
        breakBonusList.add(Material.POINTED_DRIPSTONE)
        breakBonusList.add(Material.POINTED_DRIPSTONE)
        breakBonusList.add(Material.GRAVEL)
        breakBonusList.add(Material.CLAY)
        breakBonusList.add(Material.SOUL_SAND)
        breakBonusList.add(Material.SOUL_SOIL)
        breakBonusList.add(Material.END_STONE)
        breakBonusList.add(Material.CHORUS_FLOWER)
        breakBonusList.add(Material.CHORUS_PLANT)
        breakBonusList.add(Material.PRISMARINE)
    }

    /**
     * プレイヤーが、特別に愛護・庇護すべき動物を殴った場合、ペナルティを生じさせます。
     *
     * ### ネコを殴った場合
     * あまりにもありえない行為のため考えたくもありませんが、
     * そのような人からは100EPを没収します。
     *
     * ### 他者のペットを殴った場合
     * 動物愛護の観点だけでなく、他者の所有物に損害を及ぼしたという点でも10EPを没収します。
     * ただし、スケルトンホースはスケルトンが手懐けている場合もあるため、例外です。
     *
     * ### 子供のモブを殴った場合
     * 子供に危害を加えるのは人として最低な行為です。10EPを没収します。
     * ただし、モンスターの子供は例外です。
     */
    @EventHandler(priority = EventPriority.NORMAL)
    fun onPlayerDamageFriendlyCreatures(e: EntityDamageByEntityEvent) {
        val killer = e.damager as? Player ?: return
        val victim = e.entity as? LivingEntity ?: return

        if (victim.fromMobSpawner()) return
        if (victim.isCitizensNpc()) return

        if (victim is Cat || victim is Ocelot) {
            EbiPowerModule.tryTake(killer, 100)
            notification(killer, "可愛い可愛いネコちゃんを殴るなんて！100EPを失った。")
            killer.playSound(killer.location, Sound.ENTITY_ZOMBIE_VILLAGER_CURE, 0.7f, 0.5f)
            HintModule.achieve(killer, Hint.VIOLENCE_CAT)
        } else if (victim is Tameable && victim.isTamed && victim !is SkeletonHorse) {
            EbiPowerModule.tryTake(killer, 10)
            notification(killer, "ペットを殴るなんて！10EPを失った。")
            killer.playSound(killer.location, Sound.ENTITY_ZOMBIE_VILLAGER_CURE, SoundCategory.PLAYERS, 0.7f, 0.5f)
            HintModule.achieve(killer, Hint.VIOLENCE_PET)
        } else if (victim is Ageable && victim !is Monster && victim !is Hoglin && !victim.isAdult) {
            EbiPowerModule.tryTake(killer, 10)
            notification(killer, "子供を殴るなんて！10EPを失った。")
            killer.playSound(killer.location, Sound.ENTITY_ZOMBIE_VILLAGER_CURE, SoundCategory.PLAYERS, 0.7f, 0.5f)
            HintModule.achieve(killer, Hint.VIOLENCE_CHILD)
        }
    }

    /**
     * プレイヤーがモブを倒した時にエビパワーを付与します。
     */
    @EventHandler
    fun onPlayerKillMobs(e: EntityDeathEvent) {
        val victim = e.entity
        val killer = e.entity.killer ?: return

        if (playerIsInBlacklisted(killer)) return
        // don't kill cats
        if (victim is Cat || victim is Ocelot) return
        // don't kill tamed pets
        if (victim is Tameable && victim.isTamed && victim !is SkeletonHorse) return
        //don't kill non-monster children
        if (victim is Ageable && victim !is Monster && victim !is Hoglin && !victim.isAdult) return
        // ignore creatures from spawner
        if (victim.fromMobSpawner()) return
        // マグマキューブおよびスライムは大きい個体以外対象外
        if (victim is Slime && victim.size != 4) return
        // ReasonがCUSTOMかCOMMANDであれば対象外とする。他モジュール向け機能
        if (victim.entitySpawnReason == CreatureSpawnEvent.SpawnReason.CUSTOM || victim.entitySpawnReason == CreatureSpawnEvent.SpawnReason.COMMAND) return

        var ep = if ("nightmare2" == killer.world.name) EbiPowerModule.getMobDropEP(victim, e) else 6
        val buff = getMobDropBonus(killer.inventory.itemInMainHand) * 4
        ep += if (buff > 0) random.nextInt(buff) else 0
        if (ep > 0) {
            EbiPowerModule.tryGive(killer, ep)
            HintModule.achieve(killer, Hint.KILL_MOB_AND_EARN_MONEY)
        }
    }

    /**
     * 最大まで成長した作物を伐採したとき、エビパワーを付与します。
     * 付与される値は、作物1つにつき（1+幸運の数値）EPです。
     */
    @EventHandler
    fun onHarvestCrops(e: BlockBreakEvent) {
        val p = e.player
        if (playerIsInBlacklisted(p)) return
        val blockData = e.block.blockData
        if (blockData is org.bukkit.block.data.Ageable && blockData.age == blockData.maximumAge) {
            val tool = p.inventory.itemInMainHand
            val bonus = getBlockDropBonus(tool)
            val power = (1 + bonus) * HARVEST_POWER_MULTIPLIER
            EbiPowerModule.tryGive(p, power)
            // もし幸運ボーナスがあれば30%の確率で耐久が減っていく
            if (bonus > 0 && random.nextInt(100) < 30) {
                tool.editMeta {
                    if (it !is Damageable) return@editMeta
                    it.damage += 1
                }
            }
        }
    }

    /**
     * 動物を繁殖させた場合、2EP貰えます。
     */
    @EventHandler
    fun onBreedEntities(e: EntityBreedEvent) {
        val breeder = e.breeder as? Player ?: return
        if (playerIsInBlacklisted(breeder)) return
        EbiPowerModule.tryGive(breeder, 2)
        HintModule.achieve(breeder, Hint.BREED_AND_EARN_MONEY)
    }

    /**
     * ブロックを採掘した場合、ブロックの種別やツルハシの幸運値によってエビパワーを貰えます。
     */
    @EventHandler
    fun onMineBlocks(e: BlockBreakEvent) {
        if (!breakBonusList.contains(e.block.type)) return
        if (playerIsInBlacklisted(e.player)) return
        val record = PlayerStore.open(e.player)
        val brokenBlocksCount = record.getInt(EbiPowerModule.PS_KEY_BROKEN_BLOCKS_COUNT)

        if (!e.isDropItems) return

        record[EbiPowerModule.PS_KEY_BROKEN_BLOCKS_COUNT] = brokenBlocksCount + 1

        if (brokenBlocksCount + 1 == BREAK_BLOCK_BONUS_LIMIT) {
            HintModule.achieve(e.player, Hint.MINERS_DREAM)
        }

        var ep = 1
        val tool = e.player.inventory.itemInMainHand
        val luck = tool.containsEnchantment(Enchantment.LOOT_BONUS_BLOCKS)
        if (luck) {
            ep += when (e.block.type) {
                Material.COAL_ORE -> 1
                Material.DEEPSLATE_COAL_ORE -> 1
                Material.IRON_ORE -> 1
                Material.DEEPSLATE_IRON_ORE -> 1
                Material.COPPER_ORE -> 1
                Material.DEEPSLATE_COPPER_ORE -> 1
                Material.GOLD_ORE -> 4
                Material.DEEPSLATE_GOLD_ORE -> 4
                Material.REDSTONE_ORE -> 2
                Material.DEEPSLATE_REDSTONE_ORE -> 2
                Material.LAPIS_ORE -> 3
                Material.DEEPSLATE_LAPIS_ORE -> 3
                Material.DIAMOND_ORE -> 7
                Material.DEEPSLATE_DIAMOND_ORE -> 7
                Material.EMERALD_ORE -> 11
                Material.DEEPSLATE_EMERALD_ORE -> 11
                Material.DEEPSLATE -> 1
                Material.OBSIDIAN -> 3
                Material.NETHER_QUARTZ_ORE -> 1
                Material.NETHER_GOLD_ORE -> 1
                Material.ANCIENT_DEBRIS -> 19
                else -> 0
            }
        }

        EbiPowerModule.tryGive(e.player, ep)
        HintModule.achieve(e.player, Hint.MINERS_NEWBIE)
    }

    @EventHandler
    fun onNewDayToResetBrokenBlocksCount(e: RealTimeNewDayEvent) {
        PlayerStore.openAll().forEach {
            it[EbiPowerModule.PS_KEY_BROKEN_BLOCKS_COUNT] = 0
        }
    }

    @EventHandler
    fun onTransferPlayerData(e: TransferPlayerDataEvent) {
        val power = EbiPowerModule.get(e.from)
        EbiPowerModule.tryTake(e.from, power)
        EbiPowerModule.tryGive(e.to, power)
    }

    private fun notification(p: Player, mes: String) {
        p.sendActionBar(Component.text(mes))
        // p.playSound(p.getLocation(), Sound.ENTITY_PLAYER_LEVELUP, SoundCategory.PLAYERS, 1, 2);
    }

    private fun playerIsInBlacklisted(p: Player): Boolean {
        // クリエイティブモードであれば、エビパワー取得対象外とする
        return p.gameMode == GameMode.CREATIVE || !WorldModule.getWorldInfo(p.world).canEarnEbiPower
    }

    private fun getMobDropBonus(stack: ItemStack): Int {
        return stack.getEnchantmentLevel(Enchantment.LOOT_BONUS_MOBS)
    }

    private fun getBlockDropBonus(stack: ItemStack): Int {
        return stack.getEnchantmentLevel(Enchantment.LOOT_BONUS_BLOCKS)
    }

    private val random = Random()
}